backend/backend/common/room.py

```diff
@@ -0,0 +1,44 @@
+import uuid
+
+from cached_property import cached_property
+
+from .member import RoomMember
+
+from backend.errors.room import PlayerNotInRoom, PlayerIsNotLeader
+
+
+class Room:
+    def __init__(self):
+        self._members = {}
+
+    @property
+    def members(self):
+        return self._members
+
+    @property
+    def leader(self):
+        leader = next(filter(lambda player: player.leader
+                             , self._members.values()))
+        return leader
+
+    @cached_property
+    def room_id(self):
+        return uuid.uuid4().hex[0:6]
+
+    def add_member(self, member: RoomMember):
+        self._members.update({member.id: member})
+
+    def update_leader(self, player_id, new_leader_id):
+        player = self._members.get(player_id)
+        new_leader = self._members.get(new_leader_id)
+
+        if not player or not new_leader:
+            raise PlayerNotInRoom(players=[player_id, new_leader_id])
+
+        if not player.leader:
+            raise PlayerIsNotLeader(player_name=player.name)
```

Comment:
I think that if you'd simply remove this verification from the room side and have it only on the api side it would simplify the issues.
Also, I think you should have `set_leader` and `remove_leader` function and not `update` to maintain SRP to its full potential

Reply:
change the prd of this issue then - and honstly if there is an issue with one of those request a room can have a state in which it does not have a leader which I dont like

Reply:
also delete leader should not be a thibg- a room need an admin-  room without an admin is not a functional room, we is this a state we are allowing?

Reply:
It makes sense in some sort but there is no real logic here that would fail, essentially removing the need for a "transaction"

Reply:
PRD?

Reply:
product defenitions

Reply:
this will allow a state that a room will have no leader with no way to re-elect one- as someone onced said "you don't know who will press f12 and try to mess with your api"

Reply:
I meant one API call to update leader and 2 methods on the model to achieve this purpose

Reply:
the "deletion" of a leader is setting an attrbuite fron true to false- it isn't really our source of an issue
also this can create really weird states like to leaders/ no-leader depends on the order of the method calling

Reply:
I will not find it ok to write those two types of functions with verification in them- which makes this implamintation redandent

Reply:
also also I thing it's ok to spreate room-related verfication to general verfication in between the room-api and the room-common

Reply:
sure - you can resolve